2.2.20790: AddUnits#buildUnitOwnerMap:46 - IllegalStateException

## User Description
end of british move
save game - 100 % reproduceable on load

## Map
world_war_ii_v3

## Log Message
java.lang.IllegalStateException: Duplicate key b1a7076f-ea79-4dcb-b986-b27cf9dda20c (attempted merging values Italians and Italians)

## TripleA Version
2.2.20790

## Java Version
11.0.6

## Operating System
Mac OS X

## Memory
Heap utilization statistics [MB]
Used Memory: 311
Free memory: 1736
Total memory: 2048
Max memory: 2048


## Stack Trace
```
java.lang.IllegalStateException: java.lang.IllegalStateException: Duplicate key b1a7076f-ea79-4dcb-b986-b27cf9dda20c (attempted merging values Italians and Italians)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:600)
	at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:678)
	at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:737)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateParallel(ReduceOps.java:919)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at games.strategy.triplea.odds.calculator.ConcurrentBattleCalculator.calculate(ConcurrentBattleCalculator.java:234)
	at games.strategy.triplea.ai.pro.util.ProOddsCalculator.callBattleCalc(ProOddsCalculator.java:224)
	at games.strategy.triplea.ai.pro.util.ProOddsCalculator.callBattleCalc(ProOddsCalculator.java:200)
	at games.strategy.triplea.ai.pro.util.ProOddsCalculator.calculateBattleResults(ProOddsCalculator.java:133)
	at games.strategy.triplea.ai.pro.util.ProOddsCalculator.calculateBattleResults(ProOddsCalculator.java:116)
	at games.strategy.triplea.ai.pro.ProNonCombatMoveAi.moveUnitsToBestTerritories(ProNonCombatMoveAi.java:1961)
	at games.strategy.triplea.ai.pro.ProNonCombatMoveAi.doNonCombatMove(ProNonCombatMoveAi.java:152)
	at games.strategy.triplea.ai.pro.ProNonCombatMoveAi.simulateNonCombatMove(ProNonCombatMoveAi.java:69)
	at games.strategy.triplea.ai.pro.AbstractProAi.purchase(AbstractProAi.java:227)
	at games.strategy.triplea.ai.AbstractAi.start(AbstractAi.java:507)
	at games.strategy.engine.framework.ServerGame.waitForPlayerToFinishStep(ServerGame.java:537)
	at games.strategy.engine.framework.ServerGame.runStep(ServerGame.java:407)
	at games.strategy.engine.framework.ServerGame.startGame(ServerGame.java:297)
	at games.strategy.engine.framework.startup.launcher.LocalLauncher.launchInternal(LocalLauncher.java:86)
	at games.strategy.engine.framework.startup.launcher.LocalLauncher.lambda$launch$0(LocalLauncher.java:60)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalStateException: Duplicate key b1a7076f-ea79-4dcb-b986-b27cf9dda20c (attempted merging values Italians and Italians)
	at java.base/java.util.stream.Collectors.duplicateKeyException(Collectors.java:133)
	at java.base/java.util.stream.Collectors.lambda$uniqKeysMapAccumulator$1(Collectors.java:180)
	at java.base/java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at games.strategy.engine.data.changefactory.AddUnits.buildUnitOwnerMap(AddUnits.java:46)
	at games.strategy.engine.data.changefactory.AddUnits.<init>(AddUnits.java:32)
	at games.strategy.engine.data.changefactory.ChangeFactory.addUnits(ChangeFactory.java:81)
	at games.strategy.triplea.odds.calculator.BattleCalculator.calculate(BattleCalculator.java:75)
	at games.strategy.triplea.odds.calculator.ConcurrentBattleCalculator.lambda$calculate$4(ConcurrentBattleCalculator.java:222)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:952)
	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:926)
	at java.base/java.util.stream.AbstractTask.compute(AbstractTask.java:327)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)

```

Is this a regression? I thought this was fixed.

Thomas Leavitt

On Fri, Aug 14, 2020, 4:28 PM tripleabuilderbot <notifications@github.com>
wrote:

> User Description
>
> end of british move
> save game - 100 % reproduceable on load
> Map
>
> world_war_ii_v3
> Log Message
>
> java.lang.IllegalStateException: Duplicate key
> b1a7076f-ea79-4dcb-b986-b27cf9dda20c (attempted merging values Italians and
> Italians)
> TripleA Version
>
> 2.2.20790
> Java Version
>
> 11.0.6
> Operating System
>
> Mac OS X
> Memory
>
> Heap utilization statistics [MB]
> Used Memory: 311
> Free memory: 1736
> Total memory: 2048
> Max memory: 2048
> Stack Trace
>
> java.lang.IllegalStateException: java.lang.IllegalStateException: Duplicate key b1a7076f-ea79-4dcb-b986-b27cf9dda20c (attempted merging values Italians and Italians)
> 	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
> 	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
> 	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
> 	at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:600)
> 	at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:678)
> 	at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:737)
> 	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateParallel(ReduceOps.java:919)
> 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
> 	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
> 	at games.strategy.triplea.odds.calculator.ConcurrentBattleCalculator.calculate(ConcurrentBattleCalculator.java:234)
> 	at games.strategy.triplea.ai.pro.util.ProOddsCalculator.callBattleCalc(ProOddsCalculator.java:224)
> 	at games.strategy.triplea.ai.pro.util.ProOddsCalculator.callBattleCalc(ProOddsCalculator.java:200)
> 	at games.strategy.triplea.ai.pro.util.ProOddsCalculator.calculateBattleResults(ProOddsCalculator.java:133)
> 	at games.strategy.triplea.ai.pro.util.ProOddsCalculator.calculateBattleResults(ProOddsCalculator.java:116)
> 	at games.strategy.triplea.ai.pro.ProNonCombatMoveAi.moveUnitsToBestTerritories(ProNonCombatMoveAi.java:1961)
> 	at games.strategy.triplea.ai.pro.ProNonCombatMoveAi.doNonCombatMove(ProNonCombatMoveAi.java:152)
> 	at games.strategy.triplea.ai.pro.ProNonCombatMoveAi.simulateNonCombatMove(ProNonCombatMoveAi.java:69)
> 	at games.strategy.triplea.ai.pro.AbstractProAi.purchase(AbstractProAi.java:227)
> 	at games.strategy.triplea.ai.AbstractAi.start(AbstractAi.java:507)
> 	at games.strategy.engine.framework.ServerGame.waitForPlayerToFinishStep(ServerGame.java:537)
> 	at games.strategy.engine.framework.ServerGame.runStep(ServerGame.java:407)
> 	at games.strategy.engine.framework.ServerGame.startGame(ServerGame.java:297)
> 	at games.strategy.engine.framework.startup.launcher.LocalLauncher.launchInternal(LocalLauncher.java:86)
> 	at games.strategy.engine.framework.startup.launcher.LocalLauncher.lambda$launch$0(LocalLauncher.java:60)
> 	at java.base/java.lang.Thread.run(Thread.java:834)
> Caused by: java.lang.IllegalStateException: Duplicate key b1a7076f-ea79-4dcb-b986-b27cf9dda20c (attempted merging values Italians and Italians)
> 	at java.base/java.util.stream.Collectors.duplicateKeyException(Collectors.java:133)
> 	at java.base/java.util.stream.Collectors.lambda$uniqKeysMapAccumulator$1(Collectors.java:180)
> 	at java.base/java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
> 	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
> 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
> 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
> 	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
> 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
> 	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
> 	at games.strategy.engine.data.changefactory.AddUnits.buildUnitOwnerMap(AddUnits.java:46)
> 	at games.strategy.engine.data.changefactory.AddUnits.<init>(AddUnits.java:32)
> 	at games.strategy.engine.data.changefactory.ChangeFactory.addUnits(ChangeFactory.java:81)
> 	at games.strategy.triplea.odds.calculator.BattleCalculator.calculate(BattleCalculator.java:75)
> 	at games.strategy.triplea.odds.calculator.ConcurrentBattleCalculator.lambda$calculate$4(ConcurrentBattleCalculator.java:222)
> 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
> 	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
> 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
> 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
> 	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:952)
> 	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:926)
> 	at java.base/java.util.stream.AbstractTask.compute(AbstractTask.java:327)
> 	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
> 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
> 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
> 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
> 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
> 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/triplea-game/triplea/issues/7370>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABB5HAM5DEGXBG3HCJA5JD3SAXCAJANCNFSM4P76GCWA>
> .
>

--- a/game-core/src/main/java/games/strategy/triplea/ai/pro/ProNonCombatMoveAi.java
+++ b/game-core/src/main/java/games/strategy/triplea/ai/pro/ProNonCombatMoveAi.java
@@ -785,7 +785,7 @@ class ProNonCombatMoveAi {
         }
         final TreeMap<Double, Territory> estimatesMap = new TreeMap<>();
         for (final Territory t : sortedUnitMoveOptions.get(unit)) {
-          List<Unit> defendingUnits =
+          Collection<Unit> defendingUnits =
               CollectionUtils.getMatches(
                   moveMap.get(t).getAllDefenders(),
                   ProMatches.unitIsAlliedNotOwnedAir(player, data).negate());
@@ -814,7 +814,7 @@ class ProNonCombatMoveAi {
         Territory maxWinTerritory = null;
         double maxWinPercentage = -1;
         for (final Territory t : sortedUnitMoveOptions.get(unit)) {
-          List<Unit> defendingUnits =
+          Collection<Unit> defendingUnits =
               CollectionUtils.getMatches(
                   moveMap.get(t).getAllDefenders(),
                   ProMatches.unitIsAlliedNotOwnedAir(player, data).negate());
@@ -881,7 +881,7 @@ class ProNonCombatMoveAi {
               && !ProMatches.territoryHasInfraFactoryAndIsLand().test(t)) {
             continue; // skip moving air units to allied land without a factory
           }
-          List<Unit> defendingUnits =
+          Collection<Unit> defendingUnits =
               CollectionUtils.getMatches(
                   moveMap.get(t).getAllDefenders(),
                   ProMatches.unitIsAlliedNotOwnedAir(player, data).negate());
@@ -943,7 +943,7 @@ class ProNonCombatMoveAi {
           // Find current naval defense that needs transport if it isn't transporting units
           for (final Territory t : transportDefendOptions.get(transport)) {
             if (!TransportTracker.isTransporting(transport)) {
-              final List<Unit> defendingUnits = moveMap.get(t).getAllDefenders();
+              final Collection<Unit> defendingUnits = moveMap.get(t).getAllDefenders();
               if (moveMap.get(t).getBattleResult() == null) {
                 moveMap
                     .get(t)
@@ -994,7 +994,7 @@ class ProNonCombatMoveAi {
 
         // Find current land defense results for territories that unit can amphib move
         for (final Territory t : amphibMoveOptions.get(transport)) {
-          final List<Unit> defendingUnits = moveMap.get(t).getAllDefenders();
+          final Collection<Unit> defendingUnits = moveMap.get(t).getAllDefenders();
           if (moveMap.get(t).getBattleResult() == null) {
             moveMap
                 .get(t)
@@ -1055,7 +1055,7 @@ class ProNonCombatMoveAi {
                       && (moveMap.get(territoryToMoveTransport).isCanHold() || hasFactory)) {
                     final List<Unit> attackers =
                         moveMap.get(territoryToMoveTransport).getMaxEnemyUnits();
-                    final List<Unit> defenders =
+                    final Collection<Unit> defenders =
                         moveMap.get(territoryToMoveTransport).getAllDefenders();
                     defenders.add(transport);
                     final double strengthDifference =
@@ -1121,7 +1121,7 @@ class ProNonCombatMoveAi {
         final Territory t = patd.getTerritory();
 
         // Find defense result and hold value based on used defenders TUV
-        final List<Unit> defendingUnits = moveMap.get(t).getAllDefenders();
+        final Collection<Unit> defendingUnits = moveMap.get(t).getAllDefenders();
         moveMap
             .get(t)
             .setBattleResult(
@@ -1954,7 +1954,7 @@ class ProNonCombatMoveAi {
       for (final Territory t : territoriesToDefend) {
 
         // Find result with temp units
-        final List<Unit> defendingUnits = moveMap.get(t).getAllDefenders();
+        final Collection<Unit> defendingUnits = moveMap.get(t).getAllDefenders();
         moveMap
             .get(t)
             .setBattleResult(
@@ -2258,7 +2258,7 @@ class ProNonCombatMoveAi {
         }
 
         // Check to see if the territory is safe
-        final List<Unit> defendingUnits = moveMap.get(t).getAllDefenders();
+        final Collection<Unit> defendingUnits = moveMap.get(t).getAllDefenders();
         defendingUnits.add(u);
         if (moveMap.get(t).getBattleResult() == null) {
           moveMap
@@ -2392,7 +2392,7 @@ class ProNonCombatMoveAi {
           continue;
         }
         final List<Unit> attackers = moveMap.get(t).getMaxEnemyUnits();
-        final List<Unit> defenders = moveMap.get(t).getAllDefenders();
+        final Collection<Unit> defenders = moveMap.get(t).getAllDefenders();
         defenders.add(u);
         final double strengthDifference =
             ProBattleUtils.estimateStrengthDifference(proData, t, attackers, defenders);
@@ -2445,7 +2445,7 @@ class ProNonCombatMoveAi {
 
             // Check if territory is safe after all current moves
             if (moveMap.get(t).getBattleResult() == null) {
-              final List<Unit> defendingUnits = moveMap.get(t).getAllDefenders();
+              final Collection<Unit> defendingUnits = moveMap.get(t).getAllDefenders();
               moveMap
                   .get(t)
                   .setBattleResult(

--- a/game-core/src/main/java/games/strategy/triplea/ai/pro/data/ProTerritory.java
+++ b/game-core/src/main/java/games/strategy/triplea/ai/pro/data/ProTerritory.java
@@ -143,13 +143,16 @@ public class ProTerritory {
     return defenders;
   }
 
-  public Collection<Unit> getAllDefendersForCarrierCalcs(final GameData data, final GamePlayer player) {
+  public Collection<Unit> getAllDefendersForCarrierCalcs(
+      final GameData data, final GamePlayer player) {
     if (Properties.getProduceNewFightersOnOldCarriers(data)) {
       return getAllDefenders();
     }
 
     final Set<Unit> defenders =
-        new HashSet<>(CollectionUtils.getMatches(cantMoveUnits, ProMatches.unitIsOwnedCarrier(player).negate()));
+        new HashSet<>(
+            CollectionUtils.getMatches(
+                cantMoveUnits, ProMatches.unitIsOwnedCarrier(player).negate()));
     defenders.addAll(units);
     // tempUnits can already be in the units/cantMoveUnits collection
     defenders.addAll(tempUnits);

--- a/game-core/src/main/java/games/strategy/triplea/ai/pro/data/ProTerritory.java
+++ b/game-core/src/main/java/games/strategy/triplea/ai/pro/data/ProTerritory.java
@@ -11,6 +11,7 @@ import games.strategy.triplea.ai.pro.util.ProOddsCalculator;
 import games.strategy.triplea.delegate.Matches;
 import games.strategy.triplea.delegate.TransportTracker;
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
@@ -134,21 +135,23 @@ public class ProTerritory {
     maxScrambleUnits = new ArrayList<>(patd.getMaxScrambleUnits());
   }
 
-  public List<Unit> getAllDefenders() {
-    final List<Unit> defenders = new ArrayList<>(units);
+  public Collection<Unit> getAllDefenders() {
+    final Set<Unit> defenders = new HashSet<>(units);
     defenders.addAll(cantMoveUnits);
+    // tempUnits can already be in the units/cantMoveUnits collection
     defenders.addAll(tempUnits);
     return defenders;
   }
 
-  public List<Unit> getAllDefendersForCarrierCalcs(final GameData data, final GamePlayer player) {
+  public Collection<Unit> getAllDefendersForCarrierCalcs(final GameData data, final GamePlayer player) {
     if (Properties.getProduceNewFightersOnOldCarriers(data)) {
       return getAllDefenders();
     }
 
-    final List<Unit> defenders =
-        CollectionUtils.getMatches(cantMoveUnits, ProMatches.unitIsOwnedCarrier(player).negate());
+    final Set<Unit> defenders =
+        new HashSet<>(CollectionUtils.getMatches(cantMoveUnits, ProMatches.unitIsOwnedCarrier(player).negate()));
     defenders.addAll(units);
+    // tempUnits can already be in the units/cantMoveUnits collection
     defenders.addAll(tempUnits);
     return defenders;
   }

--- a/game-core/src/main/java/games/strategy/triplea/ai/pro/util/ProTransportUtils.java
+++ b/game-core/src/main/java/games/strategy/triplea/ai/pro/util/ProTransportUtils.java
@@ -314,7 +314,7 @@ public final class ProTransportUtils {
   public static boolean validateCarrierCapacity(
       final GamePlayer player,
       final Territory t,
-      final List<Unit> existingUnits,
+      final Collection<Unit> existingUnits,
       final Unit newUnit) {
     final GameData data = player.getData();
 

